packages/block-editor/src/components/use-moving-animation/index.js

```diff
@@ -56,100 +42,86 @@ function useMovingAnimation( {
-			return () => {};
-		}
-
-		const prevRect = ref.current.getBoundingClientRect();
-		return () => {
-			const blockRect = ref.current.getBoundingClientRect();
-			const diff = blockRect.top - prevRect.top;
+		function preserveScrollPosition() {
+			if ( adjustScrollingRef.current && prevRect ) {
+				const blockRect = ref.current.getBoundingClientRect();
+				const diff = blockRect.top - prevRect.top;
 
-			if ( diff ) {
-				scrollContainer.scrollTop += diff;
+				if ( diff ) {
+					scrollContainer.scrollTop += diff;
+				}
 			}
-		};
-	}, [ triggerAnimationOnChange, adjustScrolling ] );
-
-	useLayoutEffect( () => {
-		if ( triggeredAnimation ) {
-			endAnimation();
-		}
-	}, [ triggeredAnimation ] );
-	useLayoutEffect( () => {
-		if ( ! previous ) {
-			return;
 		}
 
 		if ( prefersReducedMotion ) {
 			// If the animation is disabled and the scroll needs to be adjusted,
 			// just move directly to the final scroll position.
 			preserveScrollPosition();
-
 			return;
 		}
 
+		const controller = new Controller( {
+			x: 0,
+			y: 0,
+			config: { mass: 5, tension: 2000, friction: 200 },
+			onChange( { value } ) {
+				if ( ! ref.current ) {
+					return;
+				}
+				let { x, y } = value;
+				x = Math.round( x );
+				y = Math.round( y );
+				const finishedMoving = x === 0 && y === 0;
+				ref.current.style.transformOrigin = 'center center';
+				ref.current.style.transform = finishedMoving
+					? null // Set to `null` to explicitly remove the transform.
+					: `translate3d(${ x }px,${ y }px,0)`;
+				ref.current.style.zIndex = isSelectedRef.current ? '1' : '';
+				preserveScrollPosition();
+			},
+		} );
+
 		ref.current.style.transform = undefined;
 		const destination = getAbsolutePosition( ref.current );
 
-		triggerAnimation();
-		setTransform( {
-			x: Math.round( previous.left - destination.left ),
-			y: Math.round( previous.top - destination.top ),
-		} );
-	}, [ triggerAnimationOnChange ] );
+		const x = Math.round( previous.left - destination.left );
+		const y = Math.round( previous.top - destination.top );
 
-	function onChange( { value } ) {
-		if ( ! ref.current ) {
-			return;
-		}
-		let { x, y } = value;
-		x = Math.round( x );
-		y = Math.round( y );
-		const finishedMoving = x === 0 && y === 0;
-		ref.current.style.transformOrigin = 'center center';
-		ref.current.style.transform = finishedMoving
-			? null // Set to `null` to explicitly remove the transform.
-			: `translate3d(${ x }px,${ y }px,0)`;
-		ref.current.style.zIndex = isSelected ? '1' : '';
-
-		preserveScrollPosition();
-	}
-
-	useSpring( {
-		from: {
-			x: transform.x,
-			y: transform.y,
-		},
-		to: {
-			x: 0,
-			y: 0,
-		},
-		reset: triggeredAnimation !== finishedAnimation,
-		config: { mass: 5, tension: 2000, friction: 200 },
-		immediate: prefersReducedMotion,
-		onChange,
-	} );
+		controller.start( { x: 0, y: 0, from: { x, y } } );
+
+		return () => {
+			controller.stop();
+		};
+	}, [ previous, prevRect, prefersReducedMotion ] );
```

Comment:
I think you can simplify even further :)

 - Basically this effect is what triggers the animation, so you should remove all these deps and only have `triggerAnimationOnChange` as a dependency here.
 - previous and prevRect can be computed in the beginning of the effect instead.
 - prefersReducedMotion shouldn't trigger a new animation when it changes, so it needs to be a ref like isSelectedRef and adjustScalingRef

Reply:
It needs to be calculated at render time before the React updates the DOM unfortunately.

But yes, I also found that the reduced motion query is expensive, so we shouldn't call that at render time, we can check right before the animation start.